hal/source/bmp180_device.cpp

```diff
@@ -0,0 +1,336 @@
+		const vector<uint8_t>& value) {
+	if (key != kI2CAddress)
+		return kInvalidArgument;
+
+	write_addr_ = value[0];
+	read_addr_ = value[0] | 0x01 ;
+
+	return kNoError;
+}
+
+error_condition Bmp180Device::Set(const string& key, const string& value) {
+	return kNoError;
+}
+
+error_condition Bmp180Device::Get(const string& key, unsigned int& result) {
+	if (!IsInitialized())
+		return make_error_condition(errc::no_such_device);
+
+	if (key == kTemperature) {
+		int temperature;
+		if (GetTemperature(temperature) != kNoError) {
+			return make_error_condition(errc::io_error);
+		}
+		result = static_cast<unsigned int>(temperature);
+	} else if (key == kPressure) {
+		int pressure;
+		if (GetPressure(pressure) != kNoError) {
+			return make_error_condition(errc::io_error);
+		}
+		result = static_cast<unsigned int>(pressure);
+	}
+	return kNoError;
+}
+
+error_condition Bmp180Device::Get(const string& key, vector<uint8_t>& result) {
+	return kNoError;
+}
+
+error_condition Bmp180Device::Get(const string& key, string& value) {
+	return kNoError;
+}
+
+error_condition Bmp180Device::BusRead(uint8_t reg, uint8_t len, \
+		vector<uint8_t>& read_data) {
+	vector<uint8_t> buffer{read_addr_, reg, len};
+	if (serial_.Set(SerialPort::kData, buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+
+	DelayMSec(5);
+
+	if (serial_.Get(SerialPort::kData, read_data) != kNoError)
+		return make_error_condition(errc::io_error);
+
+	return kNoError;
+}
+
+error_condition Bmp180Device::BusWrite(uint8_t reg, \
+		vector<uint8_t>& write_data) {
+	vector<uint8_t> buffer{write_addr_, reg};
+	buffer.insert(buffer.end(), write_data.begin(), write_data.end());
+	if (serial_.Set(serial_.kData, buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+
+	DelayMSec(3);
+
+	buffer.clear();
+	buffer.resize(kBUFFER_SIZE);
+	if (serial_.Get(serial_.kData, buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+
+	return kNoError;
+}
+
+error_condition Bmp180Device::GetSystemParam(void) {
+	if (!IsInitialized())
+		return make_error_condition(errc::no_such_device);
+
+	vector<uint8_t> buffer(kBUFFER_SIZE);
+	if (BusRead(kCHIP_ID_REG, kGEN_READ_WRITE_DATA_LENGTH, buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+	chip_id_ = buffer[0];
+
+	oversamp_setting_ = kINITIALIZE_OVERSAMP_SETTING;
+	sw_oversamp_ = kINITIALIZE_SW_OVERSAMP;
+
+	if (BusRead(kVERSION_REG, kGEN_READ_WRITE_DATA_LENGTH, buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+	ml_version_ = buffer[0] & 0x0F;
+	al_version_ = (buffer[0] & 0xF0) >> 4;
+
+	return kNoError;
+}
+
+error_condition Bmp180Device::GetCalibParam(void) {
+	if (!IsInitialized())
+		return make_error_condition(errc::no_such_device);
+
+	vector<uint8_t> params_buffer(kCALIB_DATA_SIZE);
+	if (BusRead(kPROM_START_ADDR, kPROM_DATA_LEN, params_buffer) != kNoError)
+		return make_error_condition(errc::io_error);
+
+	calib_param_.ac1 = static_cast<int>(\
```

Comment:
IMHO no need to do explicit casting here, if there is no warning, let the compiler handle the casting.